src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -142,47 +142,111 @@ public string FullName
+                    // Generic`1[[ArgTypeName, ThisAssemblyNamePart]]
+                    if ((currentInfo.flags & NameFlags.GenericArg) != 0)
                     {
-                        Debug.Fail("Pre-allocated full name should have been provided in the ctor");
+                        if ((currentInfo.flags & (NameFlags.GenericTypeDefinition | NameFlags.SuffixAlreadyPrinted)) == 0)
+                        {
+                            builder.Append(']'); // every generic arg ends with ']'
+                            if ((currentInfo.flags & NameFlags.LastGenericArg) != 0)
+                            {
+                                builder.Append(']'); // the last generic arg ends with ']]'
+                            }
+
+                            if (current.AssemblyName is not null) // generic arg is always fully qualified
+                            {
+                                AppendInReverseOrder(ref builder, current.AssemblyName.FullName.AsSpan());
+                                builder.Append(" ,");
+                            }
+                        }
+                    }
+
+                    if (current.IsArray || current.IsPointer || current.IsByRef)
+                    {
+                        TypeNameParserHelpers.AppendReversedRankOrModifierStringRepresentation(current._rankOrModifier, ref builder);
+                        stack.Add(new(current.GetElementType(), currentInfo.flags | NameFlags.SuffixAlreadyPrinted));
+                    }
+                    else if (current.IsSimple && current._fullName is not null)
+                    {
+                        if ((currentInfo.flags & NameFlags.GenericTypeDefinition) != 0)
+                        {
+                            builder.Append('['); // generic type definition is followed with '[': "List`1["
+                        }
+
+                        int length = current._nestedNameLength > 0 ? current._nestedNameLength : current._fullName.Length;
+                        AppendInReverseOrder(ref builder, current._fullName.AsSpan(0, length));
+
+                        if ((currentInfo.flags & NameFlags.GenericArg) != 0)
+                        {
+                            builder.Append('['); // every generic arg starts with '['
+
+                            if ((currentInfo.flags & NameFlags.FirstGenericArg) == 0)
+                            {
+                                builder.Append(','); // every generic arg except the first one is followed with ','
+                            }
+                        }
                     }
                 }
-                else if (_nestedNameLength > 0 && _fullName.Length > _nestedNameLength) // Declaring types
+
+                Span<char> characters = builder.RawChars.Slice(0, builder.Length);
+                characters.Reverse();
+                _fullName = characters.ToString();
+                builder.Dispose();
```

Comment:
`ToString` already calls `Dispose`

Reply:
But in this case I am not calling `builder.ToString()`, but `builder.RawChars.ToString` which does not call the `Dispose`